src/game/client/shadow/w3dvolumetricshadow.cpp

```diff
@@ -0,0 +1,512 @@
+/**
+ * @file
+ *
+ * @author tomsons26
+ *
+ * @brief
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "w3dvolumetricshadow.h"
+#include "hlod.h"
+#include "mesh.h"
+#include "meshmdl.h"
+#include "missing.h"
+#include <cmath>
+#include <new>
+
+class MeshModelClass;
+
+W3DShadowGeometryMesh::W3DShadowGeometryMesh() :
+    m_polyNeighbors(nullptr),
+    m_numPolyNeighbors(0),
+    m_parentVerts(nullptr),
+    m_polygonNormals(nullptr),
+    // BUGFIX Init all members
+    m_mesh(nullptr),
+    m_modelIndex(0),
+    m_verts(nullptr),
+    m_parentGeometry(nullptr),
+    m_numVerts(0),
+    m_numPolygons(0),
+    m_polygons(nullptr)
+{
+}
+
+W3DShadowGeometryMesh::~W3DShadowGeometryMesh()
+{
+    Delete_Neighbors();
+
+    if (m_parentVerts != nullptr) {
+        delete[] m_parentVerts;
+    }
+
+    if (m_polygonNormals != nullptr) {
+        delete[] m_polygonNormals;
+    }
+}
+
+PolyNeighbor *W3DShadowGeometryMesh::Get_Poly_Neighbor(int index)
+{
+    if (m_polyNeighbors == nullptr) {
+        Build_Polygon_Neighbors();
+    }
+
+    if (index < 0 || index >= m_numPolyNeighbors) {
+        captainslog_assert(0);
+        return nullptr;
+    }
+
+    return &m_polyNeighbors[index];
+}
+
+void W3DShadowGeometryMesh::Build_Polygon_Neighbors()
+{
+    Build_Polygon_Normals();
+
+    int num = Get_Num_Polygon();
+
+    if (num == 0) {
+        if (m_numPolyNeighbors != 0) {
+            Delete_Neighbors();
+        }
+        return;
+    }
+
+    if (num != m_numPolyNeighbors) {
+        Delete_Neighbors();
+
+        if (!Allocate_Neighbors(num)) {
+            return;
+        }
+    }
+
+    for (int i = 0; i < m_numPolyNeighbors; ++i) {
+
+        m_polyNeighbors[i].myIndex = (short)i;
+
+        for (int j = 0; j < 3; ++j) {
```

Comment:
you could have alternatively gone for: ```for (auto& neighbor : m_polyNeighbors[i].neighbor) neighbor.neighborIndex = -1;``` which would have avoided the magic number of 3.

Reply:
Using `ARRAY_SIZE(m_polyNeighbors[i].neighbor)` is also ok.